### DISCUSSION TITLE
CRM API Help

### DISCUSSION THREAD

**Discussion by mirmillo** [eventum/eventum#504] [Q&A]
Hey guys, like many of us, I am currently going through the motions of updating a happily working Eventum 2.2 to the latest versions for continued PHP support. (Thanks for making something that has served my team for that long!)

I've successfully migrated my database and my workflow.  I'm stuck on my old customer integration (formerly ```/includes/customer/{customintegration}.php```).

It seems this whole section of code is completely different in 3.5.6, without an ```Abstract_Backend``` class to model from and no updated doc.  So I would like to help.

Can you confirm that this is the correct thinking...I'll need to break my current custom integration into these new CRM integration classes.

my custom CRM classes go in ```/config/crm/```
I'll end up having up to 5 files in this config directory for crm, customer, contact, contract and support_level.

It's structured like this: CRM has customers.  Customers have contacts, contracts and support_levels

## CRM
The main custom CRM class should begin with 
```
class {custom}_CRM_Backend extends CRM
```
https://github.com/eventum/eventum/blob/v3.5.6/lib/eventum/class.crm.php 

My custom CRM class sets up the db connection to my production CRM system.  And glues the other custom classes together.

## Customer
```
class {custom}_Customer_Backend extends CRM/Customer
```
https://github.com/eventum/eventum/blob/v3.5.6/lib/eventum/crm/class.customer.php

## Contact
```
class {custom}_Contact_Backend extends CRM/Contact
```
https://github.com/eventum/eventum/blob/v3.5.6/lib/eventum/crm/class.contact.php

My team does not use Contracts or Support Levels, but they would be similar.

In each case, I need to take the functionality in my current customer integration (```{customintegration}.php```) and reassign it to the abstract functions in my new custom extending classes in ```/config/crm/*```.

That sound right?

**Version(s) affected**: 3.5.6

**Comment by balsdorf** [comment]
You are right on target with your thinking. It has been a while since the last time I converted a customer backend to a CRM backend but I believe most of the methods map over to the new system, just split up into separate classes.  

For support levels and contracts, I don't recall if you can totally ignore them or if you will need to have some basic code that just returns a stub contract.

Sorry we don't have good documentation for this, but feel free to ask me any questions you might have and I'll do my best to help.

**Comment by glensc** [comment]
@balsdorf re-did customer integration in 2.4, or 3.0, I think.

However, While the models may make sense, I don't completely agree with the integration technology. I've been migrating workflow methods to events system, last most recent integration change: https://github.com/eventum/eventum/pull/497

I believe Customer and Partner integrations should be moved to that approach as well. However as I don't use neither of them, for me to change integration is hard.
and even Workflow integration transition to events is not completed.

**Comment by mirmillo** [comment]
Hey Guys, 

I'm just now getting back to this...and I'm stuck.

Here's what I did...
I copied all the abstract classes into ```config/crm```.  I changed all the ```abstract``` functions to ```public``` and gave them comments for bodies.  So everything is a stub, just so I had a place to start from.

I removed any existing ```public``` functions that had code in them (since I get them from the extending).
My crm directory looks like this:

```
15:05 $ ls -al
total 80
drwxr-xr-x@  7 ddamstra  staff    224 Aug  1 15:05 .
drwxr-xr-x@ 15 ddamstra  staff    480 Aug  1 13:38 ..
-rw-r--r--@  1 ddamstra  staff   4744 Aug  1 14:58 class.web_contact.php
-rw-r--r--@  1 ddamstra  staff   8697 Aug  1 14:53 class.web_contract.php
-rw-r--r--@  1 ddamstra  staff  10431 Aug  1 15:03 class.web_crm.php
-rw-r--r--@  1 ddamstra  staff   3058 Aug  1 14:56 class.web_customer.php
-rw-r--r--@  1 ddamstra  staff   1017 Aug  1 14:54 class.web_support_level.php
✔ ~/Sites/eventum-3.5.6/config/crm [master|✚ 81…47] 
15:05 $ 
```

I made all my classes have fresh class names extending the abstract name, such as:

```
class Web_Contact_Backend extends Contact
class Web_Contract_Backend extends Contract
class WebCrm_CRM_Backend extends CRM
class Web_Customer_Backend extends Customer
class Web_Support_Level_Backend extends Support_Level
```

Now, when I go to Manage Projects in my browser, I don't get any Customer Integrations listed in the drop down.

<img width="554" alt="Screen Shot 2019-08-01 at 3 10 33 PM" src="https://user-images.githubusercontent.com/304147/62320587-8c043900-b46e-11e9-834a-dd8265393ac3.png">

I'm not getting any error messages in the logs, or on screen (I turned error_reporting on).

Thoughts?  Ideas?

**Comment by glensc** [comment]
so, your classes are not located.

there are two options:
1. enable legacy loader. this support may be removed in the future:
```
$ ./bin/extension.php src/Extension/BuiltinLegacyLoaderExtension.php "\Eventum\Extension\BuiltinLegacyLoaderExtension"
Enabling extension: \Eventum\Extension\BuiltinLegacyLoaderExtension
```
2. create [extension](https://github.com/eventum/eventum/blob/v3.7.2/docs/examples/extension/ExampleExtension.php) providing [CrmProvider](https://github.com/eventum/eventum/blob/v3.7.2/src/Extension/Provider/CrmProvider.php) interface. you may want to also implement [AutoloadProvider](https://github.com/eventum/eventum/blob/v3.7.2/src/Extension/Provider/AutoloadProvider.php) in your extension to load the classes. This is the future, altho it may not be complete, and could change again :)

**Comment by glensc** [comment]
for legacy loader, you need to make your filename and class name match:

`class.web_contact.php` must contain a class `\Web_Contact` (case is irrelevant).

**Comment by mirmillo** [comment]
Thanks guys, I appreciate it.  I have the integration displaying in the drop down now.  Now to actually get it to work. 

****************

Some notes, so I can use this to improve the documentation.

1.  In setup.php, already had the LegacyLoader enabled for extensions, this was how WorkFlows was working.
2. The classes need to match the filenames, as @glensc mentioned
3. but the workflow class, it seems less restrictive...my workflow class is named like this ```class WebServices_Workflow_Backend extends Abstract_Workflow_Backend``` with a filename of ```class.webservices.php```.
4. For the CRM portion, I had to remove _Backend from all the class names so they ended up like this...
```
✔ ~/Sites/eventum-3.5.6/config/crm [master|✚ 81…47] 
08:02 $ ls -al
total 80
drwxr-xr-x@  7 ddamstra  staff    224 Aug  2 08:02 .
drwxr-xr-x@ 15 ddamstra  staff    480 Aug  1 17:08 ..
-rw-r--r--@  1 ddamstra  staff   4735 Aug  2 08:01 class.webcontact.php
-rw-r--r--@  1 ddamstra  staff   8688 Aug  2 08:01 class.webcontract.php
-rw-r--r--@  1 ddamstra  staff  10422 Aug  2 08:01 class.webcrm.php
-rw-r--r--@  1 ddamstra  staff   3049 Aug  2 08:01 class.webcustomer.php
-rw-r--r--@  1 ddamstra  staff   1007 Aug  2 08:02 class.websupportlevel.php
```
and class names like this

```
class WebContact extends Contact
class WebContract extends Contract
class WebCRM extends CRM
class WebCustomer extends Customer
class WebSupportLevel extends Support_Level
```

I've saved these as sanitized examples.

This is for version 3.5.6, and the legacy loader will probably be [deprecated in the future](https://github.com/eventum/eventum/issues/504#issuecomment-517469055).

**Comment by glensc** [comment]
some clarifications:

for CRM only restrictions using legacy loader are:
- does class name is irrelevant as long as it `extends CRM`, could be even `SevenMonkeys`
- class name must match filename: `class.<CLASSNAME>.php` (case insensitive)

see: https://github.com/eventum/eventum/blob/v3.7.2/lib/eventum/class.crm.php#L750

for workflow, class must end with `_Workflow_Backend`:
- https://github.com/eventum/eventum/blob/v3.7.2/lib/eventum/class.workflow.php#L1062

the legacy loader is already deprecated since [3.2.0](https://github.com/eventum/eventum/releases/tag/v3.2.0) (This version adds replacement for backend classes with adding Extension support, see #259), what my note said that it may be removed in the future (i.e if maintaining that becomes impossible).

**Comment by glensc** [comment]
Wiki has been imported to this repository (#644, #645). Pull Request to improve CRM API related documentation is very welcome (or any kind of documentation update).
